flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java

```diff
@@ -0,0 +1,428 @@
+		final TestingResourceManagerDriver.Builder driverBuilder = new TestingResourceManagerDriver.Builder();
+		final TestingSlotManagerBuilder slotManagerBuilder = new TestingSlotManagerBuilder();
+
+		private ActiveResourceManager<ResourceID> resourceManager;
+		private TestingFatalErrorHandler fatalErrorHandler;
+
+		ActiveResourceManager<ResourceID> getResourceManager() {
+			return resourceManager;
+		}
+
+		TestingFatalErrorHandler getFatalErrorHandler() {
+			return fatalErrorHandler;
+		}
+
+		void runTest(RunnableWithException testMethod) throws Exception {
+			fatalErrorHandler = new TestingFatalErrorHandler();
+			resourceManager = createAndStartResourceManager(
+					flinkConfig,
+					driverBuilder.build(),
+					slotManagerBuilder.createSlotManager());
+
+			try {
+				testMethod.run();
+			} finally {
+				resourceManager.close();
+			}
+		}
+
+		private ActiveResourceManager<ResourceID> createAndStartResourceManager(
+				Configuration configuration,
+				ResourceManagerDriver<ResourceID> driver,
+				SlotManager slotManager) throws Exception {
+			final TestingRpcService rpcService = new TestingRpcService(configuration);
+			final MockResourceManagerRuntimeServices rmServices = new MockResourceManagerRuntimeServices(rpcService, TIMEOUT_TIME, slotManager);
+
+			final ActiveResourceManager<ResourceID> activeResourceManager = new ActiveResourceManager<>(
+					driver,
+					configuration,
+					rpcService,
+					ResourceID.generate(),
+					rmServices.highAvailabilityServices,
+					rmServices.heartbeatServices,
+					rmServices.slotManager,
+					NoOpResourceManagerPartitionTracker::get,
+					rmServices.jobLeaderIdService,
+					new ClusterInformation("localhost", 1234),
+					fatalErrorHandler,
+					UnregisteredMetricGroups.createUnregisteredResourceManagerMetricGroup());
+
+			activeResourceManager.start();
+			rmServices.grantLeadership();
+
+			return activeResourceManager;
+		}
+
+		public void runInMainThread(Runnable runnable) {
+			resourceManager.handleInMainThread(runnable);
+		}
+
+		public <T> CompletableFuture<T> runInMainThread(Callable<T> callable) {
+			return resourceManager.runInMainThread(callable, TIMEOUT_TIME);
+		}
+
+		CompletableFuture<RegistrationResponse> registerTaskExecutor(ResourceID resourceID) throws Exception {
+			final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+					.createTestingTaskExecutorGateway();
+			((TestingRpcService) resourceManager.getRpcService()).registerGateway(resourceID.toString(), taskExecutorGateway);
+
+			final TaskExecutorRegistration taskExecutorRegistration = new TaskExecutorRegistration(
+					resourceID.toString(),
+					resourceID,
+					1234,
+					new HardwareDescription(1, 2L, 3L, 4L),
+					ResourceProfile.ZERO,
+					ResourceProfile.ZERO);
+
+			return runInMainThread(() -> resourceManager.registerTaskExecutor(taskExecutorRegistration, TIMEOUT_TIME))
+					.get(TIMEOUT_SEC, TimeUnit.SECONDS);
```

Comment:
We could avoid running something in the main thread by using the `ResourceManagerGateway.registerTaskExecutor` here.